### ISSUE TITLE
IDE's behaviour with recursion (and other internal errors)

### ISSUE THREAD

**Issue by neonaot** [intellij-rust/intellij-rust#7724]
## Environment

* **IntelliJ Rust plugin version:** 0.4.153.4056-212
* **Rust toolchain version:** 1.53.0 (53cb7b09b 2021-06-17) x86_64-apple-darwin
* **IDE name and version:** CLion 2021.2 (CL-212.4746.93)
* **Operating system:** macOS 11.4
* **Macro expansion engine:** new
* **Name resolution engine:** new

## Problem description
There is incorrect code with a recursion definition ( E0072, currently it doesn't detected)
And then the user tries to implement methods for trait, nothing happens (or IDE may freeze, but there isn't any information about the error):

https://user-images.githubusercontent.com/49211026/130399641-981acfae-3a95-4beb-b34e-ef04d0b9c0e1.mov

## Suggestions
* Then some action raises an error (java.lang.StackOverflowError in this case) and the action did not end, IDE must show information about the error (It can be a template with lines from logs). I think, something similar can happens with other internal errors, but I cant come up with an example.
* Then IDE works with recursion, it should expand recursion until the first repeat
example for GO language (GoLand IDE):
```go
type first struct {
	a second
	b int
	c string
}

type second struct {
	b first
}

//func main() {
//	a := second{} //feel all fields recursively 
//}
func main() {
	a := second{
		b: first{
			a: second{}, //look here
			b: 0,
			c: "",
		},
	}
}
```

## Steps to reproduce
```rust
struct F{
    p: S
}

struct S{
    b: F
}

trait Tr {
    fn f();
    const K: F;
    const L: S;
}

impl Tr for (){
    //implement method (control + enter)
}
```
